### ISSUE TITLE
BUG: IO parition_on gives Input object was not a NumPy array

### ISSUE THREAD

**Issue by raybellwaves** [geopandas/dask-geopandas#167]
Tested on latest:
```
mamba create -n test_env python=3.9 --y && conda activate test_env
pip install git+https://github.com/geopandas/dask-geopandas
pip install --extra-index-url https://pypi.fury.io/arrow-nightlies/ --prefer-binary --pre pyarrow
```

Getting an `ArrowTypeError: ('Input object was not a NumPy array', 'Conversion failed for column geometry with type geometry')` when writing a parquet:

```
import dask_geopandas as dgpd
import geopandas as gpd

# Write naturalearth_lowres.parquet using geopandas
gdf = gpd.read_file(gpd.datasets.get_path("naturalearth_lowres"))
gdf.to_parquet("naturalearth_lowres.parquet")

# Read naturalearth_lowres using dask-geopandas
dgdf = dgpd.read_parquet("naturalearth_lowres.parquet")
# Write to parquet and partition on continent
dgdf.to_parquet("naturalearth_lowres_by_continent.parquet", partition_on="continent")
```
Traceback below.

I believe this works in dask without the geometry column:
```
import dask.dataframe as dd

ddf = dd.read_parquet(
    "naturalearth_lowres.parquet",
    columns=["pop_est", "continent", "name", "iso_a3", "gdp_md_est"],
    engine="pyarrow-dataset",
)
ddf.to_parquet(
    "naturalearth_lowres_by_continent_no_geom.parquet", partition_on="continent",
    engine="pyarrow",
)
```

Traceback:
```
---------------------------------------------------------------------------
ArrowTypeError                            Traceback (most recent call last)
Input In [9], in <module>
----> 1 dgpd.read_parquet("naturalearth_lowres.parquet").to_parquet(
      2     "naturalearth_lowres_by_continent.parquet", partition_on="continent"
      3 )

File ~/miniconda3/envs/main/lib/python3.9/site-packages/dask_geopandas/core.py:450, in GeoDataFrame.to_parquet(self, path, *args, **kwargs)
    447 """See dask_geopadandas.to_parquet docstring for more information"""
    448 from .io.parquet import to_parquet
--> 450 return to_parquet(self, path, *args, **kwargs)

File ~/miniconda3/envs/main/lib/python3.9/site-packages/dask/dataframe/io/parquet/core.py:843, in to_parquet(df, path, engine, compression, write_index, append, overwrite, ignore_divisions, partition_on, storage_options, custom_metadata, write_metadata_file, compute, compute_kwargs, schema, name_function, **kwargs)
    841 graph = HighLevelGraph.from_collections(meta_name, dsk, dependencies=(data_write,))
    842 if compute:
--> 843     return compute_as_if_collection(
    844         Scalar, graph, [(meta_name, 0)], **compute_kwargs
    845     )
    846 else:
    847     return Scalar(graph, meta_name, "")

File ~/miniconda3/envs/main/lib/python3.9/site-packages/dask/base.py:317, in compute_as_if_collection(cls, dsk, keys, scheduler, get, **kwargs)
    315 schedule = get_scheduler(scheduler=scheduler, cls=cls, get=get)
    316 dsk2 = optimization_function(cls)(dsk, keys, **kwargs)
--> 317 return schedule(dsk2, keys, **kwargs)

File ~/miniconda3/envs/main/lib/python3.9/site-packages/dask/threaded.py:81, in get(dsk, result, cache, num_workers, pool, **kwargs)
     78     elif isinstance(pool, multiprocessing.pool.Pool):
     79         pool = MultiprocessingPoolExecutor(pool)
---> 81 results = get_async(
     82     pool.submit,
     83     pool._max_workers,
     84     dsk,
     85     result,
     86     cache=cache,
     87     get_id=_thread_get_id,
     88     pack_exception=pack_exception,
     89     **kwargs,
     90 )
     92 # Cleanup pools associated to dead threads
     93 with pools_lock:

File ~/miniconda3/envs/main/lib/python3.9/site-packages/dask/local.py:506, in get_async(submit, num_workers, dsk, result, cache, get_id, rerun_exceptions_locally, pack_exception, raise_exception, callbacks, dumps, loads, chunksize, **kwargs)
    504         _execute_task(task, data)  # Re-execute locally
    505     else:
--> 506         raise_exception(exc, tb)
    507 res, worker_id = loads(res_info)
    508 state["cache"][key] = res

File ~/miniconda3/envs/main/lib/python3.9/site-packages/dask/local.py:314, in reraise(exc, tb)
    312 if exc.__traceback__ is not tb:
    313     raise exc.with_traceback(tb)
--> 314 raise exc

File ~/miniconda3/envs/main/lib/python3.9/site-packages/dask/local.py:219, in execute_task(key, task_info, dumps, loads, get_id, pack_exception)
    217 try:
    218     task, data = loads(task_info)
--> 219     result = _execute_task(task, data)
    220     id = get_id()
    221     result = dumps((result, id))

File ~/miniconda3/envs/main/lib/python3.9/site-packages/dask/core.py:119, in _execute_task(arg, cache, dsk)
    115     func, args = arg[0], arg[1:]
    116     # Note: Don't assign the subtask results to a variable. numpy detects
    117     # temporaries by their reference count and can execute certain
    118     # operations in-place.
--> 119     return func(*(_execute_task(a, cache) for a in args))
    120 elif not ishashable(arg):
    121     return arg

File ~/miniconda3/envs/main/lib/python3.9/site-packages/dask/optimization.py:969, in SubgraphCallable.__call__(self, *args)
    967 if not len(args) == len(self.inkeys):
    968     raise ValueError("Expected %d args, got %d" % (len(self.inkeys), len(args)))
--> 969 return core.get(self.dsk, self.outkey, dict(zip(self.inkeys, args)))

File ~/miniconda3/envs/main/lib/python3.9/site-packages/dask/core.py:149, in get(dsk, out, cache)
    147 for key in toposort(dsk):
    148     task = dsk[key]
--> 149     result = _execute_task(task, cache)
    150     cache[key] = result
    151 result = _execute_task(out, cache)

File ~/miniconda3/envs/main/lib/python3.9/site-packages/dask/core.py:119, in _execute_task(arg, cache, dsk)
    115     func, args = arg[0], arg[1:]
    116     # Note: Don't assign the subtask results to a variable. numpy detects
    117     # temporaries by their reference count and can execute certain
    118     # operations in-place.
--> 119     return func(*(_execute_task(a, cache) for a in args))
    120 elif not ishashable(arg):
    121     return arg

File ~/miniconda3/envs/main/lib/python3.9/site-packages/dask/dataframe/io/parquet/core.py:157, in ToParquetFunctionWrapper.__call__(self, df, block_index)
    150 filename = (
    151     f"part.{part_i + self.i_offset}.parquet"
    152     if self.name_function is None
    153     else self.name_function(part_i + self.i_offset)
    154 )
    156 # Write out data
--> 157 return self.engine.write_partition(
    158     df,
    159     self.path,
    160     self.fs,
    161     filename,
    162     self.partition_on,
    163     self.write_metadata_file,
    164     **(dict(self.kwargs_pass, head=True) if part_i == 0 else self.kwargs_pass),
    165 )

File ~/miniconda3/envs/main/lib/python3.9/site-packages/dask/dataframe/io/parquet/arrow.py:726, in ArrowDatasetEngine.write_partition(cls, df, path, fs, filename, partition_on, return_metadata, fmd, compression, index_cols, schema, head, custom_metadata, **kwargs)
    723     t = t.replace_schema_metadata(metadata=_md)
    725 if partition_on:
--> 726     md_list = _write_partitioned(
    727         t,
    728         df,
    729         path,
    730         filename,
    731         partition_on,
    732         fs,
    733         preserve_index,
    734         index_cols=index_cols,
    735         compression=compression,
    736         **kwargs,
    737     )
    738     if md_list:
    739         _meta = md_list[0]

File ~/miniconda3/envs/main/lib/python3.9/site-packages/dask/dataframe/io/parquet/arrow.py:123, in _write_partitioned(table, df, root_path, filename, partition_cols, fs, preserve_index, index_cols, **kwargs)
    119     keys = (keys,)
    120 subdir = fs.sep.join(
    121     [f"{name}={val}" for name, val in zip(partition_cols, keys)]
    122 )
--> 123 subtable = pa.Table.from_pandas(
    124     subgroup,
    125     nthreads=1,
    126     preserve_index=preserve_index,
    127     schema=subschema,
    128     safe=False,
    129 )
    130 prefix = fs.sep.join([root_path, subdir])
    131 fs.mkdirs(prefix, exist_ok=True)

File ~/miniconda3/envs/main/lib/python3.9/site-packages/pyarrow/table.pxi:1782, in pyarrow.lib.Table.from_pandas()

File ~/miniconda3/envs/main/lib/python3.9/site-packages/pyarrow/pandas_compat.py:594, in dataframe_to_arrays(df, schema, preserve_index, nthreads, columns, safe)
    589     return (isinstance(arr, np.ndarray) and
    590             arr.flags.contiguous and
    591             issubclass(arr.dtype.type, np.integer))
    593 if nthreads == 1:
--> 594     arrays = [convert_column(c, f)
    595               for c, f in zip(columns_to_convert, convert_fields)]
    596 else:
    597     arrays = []

File ~/miniconda3/envs/main/lib/python3.9/site-packages/pyarrow/pandas_compat.py:594, in <listcomp>(.0)
    589     return (isinstance(arr, np.ndarray) and
    590             arr.flags.contiguous and
    591             issubclass(arr.dtype.type, np.integer))
    593 if nthreads == 1:
--> 594     arrays = [convert_column(c, f)
    595               for c, f in zip(columns_to_convert, convert_fields)]
    596 else:
    597     arrays = []

File ~/miniconda3/envs/main/lib/python3.9/site-packages/pyarrow/pandas_compat.py:581, in dataframe_to_arrays.<locals>.convert_column(col, field)
    576 except (pa.ArrowInvalid,
    577         pa.ArrowNotImplementedError,
    578         pa.ArrowTypeError) as e:
    579     e.args += ("Conversion failed for column {!s} with type {!s}"
    580                .format(col.name, col.dtype),)
--> 581     raise e
    582 if not field_nullable and result.null_count > 0:
    583     raise ValueError("Field {} was non-nullable but pandas column "
    584                      "had {} null values".format(str(field),
    585                                                  result.null_count))

File ~/miniconda3/envs/main/lib/python3.9/site-packages/pyarrow/pandas_compat.py:575, in dataframe_to_arrays.<locals>.convert_column(col, field)
    572     type_ = field.type
    574 try:
--> 575     result = pa.array(col, type=type_, from_pandas=True, safe=safe)
    576 except (pa.ArrowInvalid,
    577         pa.ArrowNotImplementedError,
    578         pa.ArrowTypeError) as e:
    579     e.args += ("Conversion failed for column {!s} with type {!s}"
    580                .format(col.name, col.dtype),)

File ~/miniconda3/envs/main/lib/python3.9/site-packages/pyarrow/array.pxi:312, in pyarrow.lib.array()

File ~/miniconda3/envs/main/lib/python3.9/site-packages/pyarrow/array.pxi:83, in pyarrow.lib._ndarray_to_array()

File ~/miniconda3/envs/main/lib/python3.9/site-packages/pyarrow/error.pxi:122, in pyarrow.lib.check_status()

ArrowTypeError: ('Input object was not a NumPy array', 'Conversion failed for column geometry with type geometry')
```

**Comment by martinfleis** [comment]
Thanks for the report! Can confirm, it does not even need nightly arrow.

**Comment by jorisvandenbossche** [comment]
@raybellwaves indeed thanks for the report! This is an issue in how the geopandas GeoDataFrame gets converted to a pyarrow.Table. For writing each dask partition to a single file, we control this conversion by overriding the `_pandas_to_arrow_table` method on the dask engine. 
But it seems that for a partitioned write, this is not being used for the subset of the dataframe. This will probably be the easiest to fix on the dask side.

**Comment by jorisvandenbossche** [comment]
Opened a PR in dask to enable this: https://github.com/dask/dask/pull/8741, and also one here to fix and test it: #171

With those two changes together, this works now:

```python
import dask_geopandas
df = geopandas.read_file(geopandas.datasets.get_path("naturalearth_lowres"))
ddf = dask_geopandas.from_geopandas(df, npartitions=1)

ddf.to_parquet("naturalearth_lowres_by_continent.parquet", partition_on="continent")
```

```
In [5]: dask_geopandas.read_parquet("naturalearth_lowres_by_continent.parquet")
Out[5]: 
Dask GeoDataFrame Structure:
              pop_est    name  iso_a3 gdp_md_est  geometry        continent
npartitions=8                                                              
                int64  object  object    float64  geometry  category[known]
                  ...     ...     ...        ...       ...              ...
...               ...     ...     ...        ...       ...              ...
                  ...     ...     ...        ...       ...              ...
                  ...     ...     ...        ...       ...              ...
Dask Name: read-parquet, 8 tasks

In [6]: list((pathlib.Path(".") / "naturalearth_lowres_by_continent.parquet").iterdir())
Out[6]: 
[PosixPath('naturalearth_lowres_by_continent.parquet/continent=Europe'),
 PosixPath('naturalearth_lowres_by_continent.parquet/continent=South America'),
 PosixPath('naturalearth_lowres_by_continent.parquet/_metadata'),
 PosixPath('naturalearth_lowres_by_continent.parquet/_common_metadata'),
 PosixPath('naturalearth_lowres_by_continent.parquet/continent=Seven seas (open ocean)'),
 PosixPath('naturalearth_lowres_by_continent.parquet/continent=Asia'),
 PosixPath('naturalearth_lowres_by_continent.parquet/continent=North America'),
 PosixPath('naturalearth_lowres_by_continent.parquet/continent=Africa'),
 PosixPath('naturalearth_lowres_by_continent.parquet/continent=Antarctica'),
 PosixPath('naturalearth_lowres_by_continent.parquet/continent=Oceania')]
```
